### ISSUE TITLE
`CupyBackend` matrices are `np.ndarray`s

### ISSUE THREAD

**Issue by BrunoLiegiBastonLiegi** [qiboteam/qibojit#196]
While working on qiboteam/qibo#1548 I noticed that if you try to get the matrices from the `CupyBackend`, namely:
``` python
from qibojit.backends import CupyBackend

b = CupyBackend()
print(type(b.matrices.X))
```
you get a `np.ndarray` rather than a `cp.ndarray`. This does not seem to be the case for other backends that do not have `np.ndarray` as native arrays, e.g. `PyTorchBackend` or `Tensorflowbackend`. 
I was wondering what was the reason behind this, because this makes some things in qiboteam/qibo#1548 more annoying, requiring for you to explicitely cast to cupy arrays when needed.

**Comment by andrea-pasquale** [comment]
I believe that there were some discussion about this in the past https://github.com/qiboteam/qibo/issues/928
In the end we decided to use numpy for all backends, which do not require other types due to automatic differentiation (such as tensorflow) since numpy is expected to be faster for small matrices https://github.com/qiboteam/qibojit/pull/125#issuecomment-1605363088
Even though I agree with the fact that it is weird to have `CupyMatrices` which are np.arrays...

**Comment by BrunoLiegiBastonLiegi** [comment]
Mmmh I see, I would still argue that at some point you have to cast to `cp.arrays` anyway, thus even if `numpy` is faster for the small matrices you will have to move to `cupy` eventually. At that point, isn't it better to just have them as `cupy` cached arrays? It is gonna be slow the first time, but identical to `numpy` afterwards.

**Comment by renatomello** [comment]
> I believe that there were some discussion about this in the past [qiboteam/qibo#928](https://github.com/qiboteam/qibo/issues/928) In the end we decided to use numpy for all backends, which do not require other types due to automatic differentiation (such as tensorflow) since numpy is expected to be faster for small matrices [#125 (comment)](https://github.com/qiboteam/qibojit/pull/125#issuecomment-1605363088) Even though I agree with the fact that it is weird to have `CupyMatrices` which are np.arrays...

This was true at the time but it changed since the `torch` inclusion because `torch` does not do autograd on `numpy` arrays.

**Comment by andrea-pasquale** [comment]
Indeed, I would say that in general we should cast to which ever type is necessary as long as we don't trigger the qiboteam/qibo#928 (or we find an alternative solution for it).
